### PR TITLE
Fix way center calculation

### PR DESCRIPTION
**Pull request by alexey-zakharenkov**
Two changes:

- More accurate center calculation for closed ways. Now a rectangular platform get its center at geometrical center.
- Some refactoring: main() function created to eliminate globals; two pieces of code converted to separate functions.

### CODE REVIEW

File: process_subways.py [hunk 1]
```diff
@@ -175,7 +183,42 @@ def calculate_relation_center(el):
...
+    """Validate cities. Return list of good cities."""
+    good_cities = []
+    for c in cities:
+        try:
+            c.extract_routes()
+        except CriticalValidationError as e:
+            logging.error(
+                "Critical validation error while processing %s: %s",
+                c.name,
+                str(e),
```
Comment (Zisam) [inline process_subways.py:203]: NIT: Tracing to `str` is not necessary

File: process_subways.py [hunk 2]
```diff
@@ -94,8 +94,16 @@ def calculate_way_center(el):
             return
         center = [0, 0]
         count = 0
-        for nd in el['nodes']:
+        way_nodes = el['nodes']
+        for i, nd in enumerate(way_nodes):
             if nd in nodes:
```
Comment (Zisam) [inline process_subways.py:99]: NIT:
```python
if nd not in nodes:
   continue
```

File: process_subways.py [hunk 3]
```diff
@@ -94,8 +94,16 @@ def calculate_way_center(el):
...
+                # Don't count the first node of a closed way twice
+                if (
+                        len(way_nodes) > 1
+                        and i == len(way_nodes) - 1
```
Comment (Zisam) [inline process_subways.py:103]: NIT: put `len(way_nodes` as a variable earlier. And check, whether i is the last index at the first place, then check other conditions.

File: process_subways.py [hunk 4]
```diff
@@ -175,7 +183,42 @@ def calculate_relation_center(el):
...
+                str(e),
+            )
+            c.error(str(e))
+        except AssertionError as e:
+            logging.error(
+                "Validation logic error while processing %s: %s",
+                c.name,
+                str(e),
+            )
+            c.error("Validation logic error: {}".format(str(e)))
```
Comment (Zisam) [inline process_subways.py:212]:
```suggestion
            c.error("Validation logic error: {e}")
```

File: process_subways.py (hunk 1 above)
Comment (alexey-zakharenkov) [reply process_subways.py]: \+

File: process_subways.py (hunk 2 above)
Comment (alexey-zakharenkov) [reply process_subways.py]: \+

File: process_subways.py (hunk 3 above)
Comment (alexey-zakharenkov) [reply process_subways.py]: \+

File: process_subways.py (hunk 4 above)
Comment (alexey-zakharenkov) [reply process_subways.py]: \+

File: process_subways.py [hunk 5]
```diff
@@ -175,7 +183,42 @@ def calculate_relation_center(el):
...
+def add_osm_elements_to_cities(osm_elements, cities):
+    for el in osm_elements:
+        for c in cities:
+            if c.contains(el):
+                c.add(el)
+
+
+def validate_cities(cities):
+    """Validate cities. Return list of good cities."""
+    good_cities = []
```
[the remark this answers is not in the document]
Comment (alexey-zakharenkov) [reply process_subways.py:197]: I would not change this. Firstly, "good/bad" has been firmly rooted in variable, method and even command-line argument names. Secondly, does "validated" mean "has been subject to validation process" or "validated successfully"?

File: process_subways.py [hunk 6]
```diff
@@ -413,3 +434,7 @@ def calculate_relation_center(el):
 
         filename = getattr(options, option_name)
         processor.process(cities, transfers, filename, options.cache)
+
+
+if __name__ == '__main__':
+    main()
```
Comment (KarinaRomanVK) [inline process_subways.py:440]: No new line at the end of file

Comment (alexey-zakharenkov) [reply process_subways.py]: \+

**Review by KarinaRomanVK — APPROVED** [review]

File: process_subways.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zisam) [reply process_subways.py:197]: I see the point. Okay!

**Pull request merged**